chore(website): coordinator docs and blog (#2757)
privacy-ethereum/maci: apps/website/blog/2025-09-05-maci-coordinator-service.md

---
slug: maci-coordinator-service
title: MACI Coordinator Service
description: Announcing our coordinator service for MACI
authors:
  - name: ctrlc03
    title: MACI former team lead
    url: https://x.com/ctrlc03
    image_url: https://avatars.githubusercontent.com/u/93448202?v=4c
  - name: NicoSerranoP
    title: MACI team member
    url: https://x.com/NicoSerranoP
    image_url: https://avatars.githubusercontent.com/u/38594836?v=4
tags: [voting, security, anonymity, MACI, coordinator]

> excerpt: "A MACI plugin for Aragon OSx to enable private voting in DAOs"

Reviewer: ```suggestion
excerpt: "Announcing the MACI Coordinator Service for automated deployment and finalization"
```

**Fear no more - the coordinator service is here to make your life easier.**

## The vision: MACI for everyone

Our goal is ambitious, yet simple: **completely eliminate the need for any technical knowledge to use MACI**. No more terminal commands, no more manual proof generation, no more complex scripts. Just configure once, run inside a Docker container on your server, and enjoy the full power of MACI's privacy-preserving voting infrastructure.

## Technical implementation

The service was built with the following technologies for reliability and scalability:

- TypeScript for type-safe, maintainable code
- Nest.js framework for enterprise-grade architecture
- REST API for standard CRUD operations
- WebSocket endpoints for real-time updates during intensive computations
- MACI SDK package to have all the logic and functions ready to use
- Docker-ready for simple deployment anywhere

The core functionalities of the service are as follows:

> - Deploy a MACI [subgraph](https://github.com/privacy-scaling-explorations/maci/tree/dev/apps/coordinator/ts/subgraph) to query on-chain data easily
> - [Deploy](https://github.com/privacy-scaling-explorations/maci/tree/dev/apps/coordinator/ts/deployer) the MACI smart contracts, including polls
> - [Finalise](https://github.com/privacy-scaling-explorations/maci/tree/dev/apps/coordinator/ts/proof) polls by processing votes and tallying the results

Reviewer: ```suggestion
- Deploy a MACI [subgraph](https://github.com/privacy-scaling-explorations/maci/tree/main/apps/coordinator/ts/subgraph) to query on-chain data easily
- [Deploy](https://github.com/privacy-scaling-explorations/maci/tree/main/apps/coordinator/ts/deployer) the MACI smart contracts, including polls
- [Finalise](https://github.com/privacy-scaling-explorations/maci/tree/main/apps/coordinator/ts/proof) polls by processing votes and tallying the results
```

- [Schedule](https://github.com/privacy-scaling-explorations/maci/tree/main/apps/coordinator/ts/scheduler) polls for automatic finalization when the voting period ends
- [Health](https://github.com/privacy-scaling-explorations/maci/tree/main/apps/coordinator/ts/health) endpoint to check everything is setup and working

Let's dive into more details in the following sections.

> ### Subraph deployment

Reviewer: ```suggestion
### Subgraph deployment
```

We chose to separate each action into its own endpoint to avoid long-running computations that could lead to network errors or timeouts.

### Schedule

After finishing up the finalisation module, we realized that it would be a lot easier for users to create polls and schedule the finalization process. This removes the need for users to manually call the finalization endpoints described above.

1. To schedule a deployed poll you can use `https://coordinator.maci.vote/v1/scheduler/register`.
2. To check if a poll is scheduled you can use `https://coordinator.maci.vote/v1/scheduler/status`
3. To delete a scheduled poll you can use `https://coordinator.maci.vote/v1/scheduler/delete`

## How to use

### Run it from source

1. Clone the MACI repository: [https://github.com/privacy-scaling-explorations/maci](https://github.com/privacy-scaling-explorations/maci)
2. Install dependencies and build the project

```bash
pnpm install
pnpm run build
```

3. Download the zkeys for the coordinator

```bash

> pnpm run download-zkeys:ceremony:coordinator

Reviewer: Does this command exist?

```bash
cd apps/coordinator
```

6. Configure your `.env` file. Make sure to configure it with your own secure values. The coordinator MACI private key from the previous step should be set here

```bash
cp .env.example .env
```

7. Start the service

```bash
pnpm run start
# or
pnpm run start:prod
```

### Run it with Docker

1. Clone the MACI repository: [https://github.com/privacy-scaling-explorations/maci](https://github.com/privacy-scaling-explorations/maci)
2. Go inside the project directory

```bash
cd maci
```

3. Build the docker image and run the container

```bash
# Build docker
docker compose -f apps/coordinator/docker-compose.yml build

# Run container detached
docker compose -f apps/coordinator/docker-compose.yml up -d
```

## Usage notes

1. You can check out the documentation while the service is running in [http://localhost:3001/api](http://localhost:3001/api)

2. We have a [e2e.deploy.test.ts](https://github.com/privacy-scaling-explorations/maci/blob/main/apps/coordinator/tests/e2e.deploy.test.ts) file that performs a complete process: deployment, voting and finalization. You can use it to guide yourself.

## Future work

> We are researching about different cryptographic methods to decentralize the coordinator. The idea is to eliminate reliance on a single centralized service for finalizing polls. A group of people would be able to spin up their own coordinator services and finalize the poll in a multi-party computation. For more info check out our MACI v4 research.

Reviewer: I'm not sure the finalisation would necessarily be a multiparty computation like that, I kept it generic - let me know if you disagree. Also we don't have anything to point to yet for v4 so removed that sentence

```suggestion
We are researching about different cryptographic methods to decentralize the coordinator. The idea is to eliminate reliance on a single centralized entity for running MACI polls. Multiple entities would be able to spin up their own coordinator services and contribute to running a MACI poll in a decentralized way.
```
